File: src/main/java/todayquest/quest/entity/DetailQuest.java
package todayquest.quest.entity;

import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import todayquest.quest.dto.DetailQuestRequestDto;

import javax.persistence.*;

import static javax.persistence.EnumType.STRING;
import static javax.persistence.GenerationType.IDENTITY;
import static lombok.AccessLevel.PROTECTED;

@Getter
@NoArgsConstructor(access = PROTECTED)
@Entity(name = "detail_quest")
public class DetailQuest {

    @Id @GeneratedValue(strategy = IDENTITY)
    @Column(name = "detail_quest_id")
    private Long id;

    @Column(nullable = false, length = 100)
    private String title;

    @Column(nullable = false)
    private Short targetCount;

    @Column(nullable = false)
    private Short count;

    @Enumerated(STRING)
    @Column(nullable = false)
    private DetailQuestType type;

    @Enumerated(STRING)
    @Column(nullable = false)
    private DetailQuestState state;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "quest_id")
    private Quest quest;

    @Builder
    public DetailQuest(Long id, String title, Short targetCount, DetailQuestType type, DetailQuestState state, Quest quest, Short count) {
        this.id = id;
        this.title = title;
        this.targetCount = targetCount;
        this.type = type;
        this.state = state;
        this.quest = quest;
        this.count = count;
    }

    public void updateDetailQuest(DetailQuestRequestDto newDetailQuest) {
        this.title = newDetailQuest.getTitle();
        this.type = newDetailQuest.getType();
        this.targetCount = newDetailQuest.getType() == DetailQuestType.COUNT ? newDetailQuest.getCount() : 1;
        resetCount();
    }

    public void resetCount() {
        this.count = 0;
    }

    public void changeState(DetailQuestState state) {
        this.state = state;
    }

    public void addCount() {
        this.count++;
        if(count.equals(targetCount)) changeState(DetailQuestState.COMPLETE);
    }
}

File: src/main/java/todayquest/quest/entity/DetailQuest.kt
package todayquest.quest.entity

import todayquest.quest.dto.DetailQuestRequestDto
import jakarta.persistence.*
import jakarta.persistence.EnumType.*
import jakarta.persistence.FetchType.*
import jakarta.persistence.GenerationType.IDENTITY

@Entity
@Table(name = "detail_quest")
class DetailQuest(
    title: String,
    targetCount: Short,
    type: DetailQuestType,
    state: DetailQuestState,
    quest: Quest,
) {

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "detail_quest_id")
    val id: Long? = null

    @Column(nullable = false, length = 100)
    var title: String = title
        protected set

    @Column(nullable = false)
    var targetCount: Short = targetCount
        protected set

    @Column(nullable = false)
    var count: Short = 0
        protected set

    @Enumerated(STRING)
    @Column(nullable = false)
    var type: DetailQuestType = type
        protected set

    @Enumerated(STRING)
    @Column(nullable = false)
    var state: DetailQuestState = state
        protected set

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "quest_id")
    val quest: Quest = quest

    fun updateDetailQuest(newDetailQuest: DetailQuestRequestDto) {
        title = newDetailQuest.title!!
        type = newDetailQuest.type!!
        targetCount = if (newDetailQuest.type == DetailQuestType.COUNT) newDetailQuest.targetCount!! else 1

        if(newDetailQuest.id != this.id || newDetailQuest.type != this.type) resetCount()

        if(count < targetCount) state = DetailQuestState.PROCEED
        else {
            state = DetailQuestState.COMPLETE
            count = targetCount
        }

    }

    fun resetCount() {
        count = 0
        state = DetailQuestState.PROCEED
    }

    fun changeState(state: DetailQuestState?) {
        this.state = state!!
    }

    fun addCount() {
        count++
        if (count == targetCount) changeState(DetailQuestState.COMPLETE)
    }

    fun changeCount(count: Short) {
        this.count = if(count > targetCount) targetCount else count

        when {
            count < targetCount -> changeState(DetailQuestState.PROCEED)
            else -> changeState(DetailQuestState.COMPLETE)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as DetailQuest

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }

}